paddle/fluid/pir/transforms/onednn/cpu_bfloat16_placement_pass.cc

```diff
@@ -0,0 +1,594 @@
+      return false;
+    }
+
+    std::string target_op_name = op->name();
+    auto op_info =
+        pir::IrContext::Instance()->GetRegisteredOpInfo(target_op_name);
+    if (op_info) {
+      std::vector<pir::Type> op_item_inner_output_types;
+      for (size_t i = 0; i < op->num_results(); ++i) {
+        op_item_inner_output_types.push_back(op->result_type(i));
+      }
+      auto attributes = op->attributes();
+
+      if (attributes.find("mkldnn_data_type") != attributes.end()) {
+        attributes["mkldnn_data_type"] =
+            pir::StrAttribute::get(pir::IrContext::Instance(), "float32");
+      }
+
+      pir::Operation *op_item_inner = rewriter.Build(op->operands_source(),
+                                                     attributes,
+                                                     op_item_inner_output_types,
+                                                     op_info);
+      rewriter.ReplaceOp(op, op_item_inner->results());
+    }
+
+    return true;
+  }
+};
+
+template <typename OpType>
+class RemoveUnsupportedOpPattern : public pir::OpRewritePattern<OpType> {
+ public:
+  using pir::OpRewritePattern<OpType>::OpRewritePattern;
+
+  bool MatchAndRewrite(
+      OpType op,
+      pir::PatternRewriter &rewriter) const override {  // NOLINT
+
+    uint32_t num_operands = op->num_operands();
+    for (uint32_t i = 0; i < num_operands; i++) {
+      auto *pre_op = pir::GetDefiningOpForInput(op, i);
+      if (pre_op->HasAttribute("mkldnn_data_type")) {
+        return false;
+      }
+    }
+
+    bool unsupported_op = false;
+    for (auto &value : op->operands_source()) {
+      pir::Type op_dtype = pir::GetDataTypeFromValue(value);
+      // Only float input can be converted to bfloat16
+      if (!op_dtype.isa<pir::Float32Type>()) {
+        unsupported_op = true;
+      }
+    }
+    if (!unsupported_op) {
+      return false;
+    }
+
+    std::string target_op_name = op->name();
+    auto op_info =
+        pir::IrContext::Instance()->GetRegisteredOpInfo(target_op_name);
+    if (op_info) {
+      std::vector<pir::Type> op_item_inner_output_types;
+      for (size_t i = 0; i < op->num_results(); ++i) {
+        op_item_inner_output_types.push_back(op->result_type(i));
+      }
+      auto attributes = op->attributes();
+      if (attributes.find("mkldnn_data_type") != attributes.end()) {
+        attributes["mkldnn_data_type"] =
+            pir::StrAttribute::get(pir::IrContext::Instance(), "float32");
+      }
+      pir::Operation *op_item_inner = rewriter.Build(op->operands_source(),
+                                                     attributes,
+                                                     op_item_inner_output_types,
+                                                     op_info);
+      rewriter.ReplaceOp(op, op_item_inner->results());
+    }
+
+    return true;
+  }
+};
+
+template <typename OpType>
+class CpuBfloat16Pattern1 : public pir::OpRewritePattern<OpType> {
```

Comment:
So compared to `OneDNNBf16PlacementPattern`, this pattern changes dtype of results instead of attr, am I right? May I know the reason why splitting these two operations into two patterns?

Reply:
This just a test pattern for add(add_) and I haven't delete yet....